File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
from collections import deque

n, m = map(int, input().split())

graph = [list(map(int, input().split())) for _ in range(n)]
result = [[-1]*m for _ in range(n)]
visited = [[0]*m for _ in range(n)]

dx = [1, -1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(x, y):
    queue = deque()
    queue.append((x,y))
    visited[x][y] = 1
    result[x][y] = 0

    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and visited[nx][ny] == 0:
                if graph[nx][ny] == 0:
                    visited[nx][ny] = 1
                    result[nx][ny] = 0
                elif graph[nx][ny] == 1:
                    visited[nx][ny] = 1
                    result[nx][ny] = result[x][y] + 1
                    queue.append((nx, ny))
    return

x = 0
y = 0
for i in range(n):
    for j in range(m):
        if graph[i][j] == 2:
            x = i
            y = j
            break

bfs(x, y)

for i in range(n):
    for j in range(m):
        if graph[i][j] == 0:
            print(0, end = " ")
        else:
            print(result[i][j], end = " ")
    print()

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

public class Main {
    static int graph[][];
    static int n;
    static int m;
    static int visited[][];
    static int dirX[] = {1, -1, 0, 0};
    static int dirY[] = {0, 0, 1, -1};
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        graph = new int[n][m];
        visited = new int[n][m];
        for(int i = 0; i < n; i++) {
            Arrays.fill(visited[i], -1);
        }
        int start[] = new int[2];
        for(int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j = 0; j < m; j++){
                int number = Integer.parseInt(st.nextToken());
                if(number == 2) {
                    start[0] = i;
                    start[1] = j;
                }
                if(number == 0) {
                    visited[i][j] = 0;
                }
                graph[i][j] = number;
            }
        }

        bfs(start[0], start[1]);
        for(int i = 0; i<n; i++) {
            for(int j = 0; j<m; j++) {
                bw.write(visited[i][j] + " ");
            }
            bw.write("\n");
        }
        bw.flush();
        bw.close();
    }
    private static void bfs(int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        visited[x][y] = 0;
        q.add(new int[] {x, y});

        while(!q.isEmpty()) {
            int[] current = q.poll();
            int dx = current[0];
            int dy = current[1];
            for(int i = 0; i < 4; i++) {
                int nx = dx + dirX[i];
                int ny = dy + dirY[i];
                if(nx >= 0 && nx < n && ny >= 0 && ny < m && visited[nx][ny] == -1 && graph[nx][ny] == 1) {
                    visited[nx][ny] = visited[dx][dy] + 1;
                    q.add(new int[] {nx, ny});
                }
            }
        }
        
    }
}
